### PR TITLE
Create the structure of a redaction processor

### PR DESCRIPTION
**Pull request by leonsp-ai**
**Description:** Implement a redaction processor to prevent leaking sensitive or private span attributes or span attribute values for privacy, security, and compliance purposes

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/4131

**Testing:** This is the first PR creating the structure of the new processor per [guidelines](https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#when-adding-a-new-component). This PR has coverage for creating, starting, and stopping the processor. Test coverage for the actual redaction functionality will go in a second PR along with the implementation.

**Documentation:** Readme for the new redaction processor

### CODE REVIEW

**Comment by leonsp-ai** [comment]
@dmitryax @Aneurysm9 Is there something I can do to make reviewing easier?

I don't see this PR on [the project board](https://github.com/orgs/open-telemetry/projects/3) yet, but I don't seem to have permissions to edit the projects field

**Comment by leonsp-ai** [comment]
@dmitryax @Aneurysm9 Do you think it would help if I streamlined this PR further?

**Comment by leonsp-ai** [comment]
Bump

**Review by mx-psi — COMMENTED** [review]
Left a few more comments. I don't think you can access `service::telemetry::logs` from the processor. I would suggest keeping the summary/dryRun functionality to a future PR

**Review by mx-psi — APPROVED** [review]
PR looks good, but the issue above needs to be fixed for this to mergeable

File: processor/redactionprocessor/processor.go
Comment (leonsp-ai) [reply processor/redactionprocessor/processor.go:92]: @dmitryax Using this processor in production, I've found that the counts are convenient for querying through traces for redaction. The lists of attributes are nice for diagnostics, but being able to query for `redactedKeysCount > 0` is convenient

The design reason for the counts is that they leak less information than the lists of span attribute keys in case the actual key names need to be kept confidential

Comment (dmitryax) [reply processor/redactionprocessor/processor.go:92]: Hey @leonsp-ai, sorry for the delayed response. Thanks for explaining the attributes purpose. It looks good to me.

Also attribute names in this PR don't fit established attribute naming conventions. Please take a look at [this specification](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/trace/semantic_conventions) and think about names that would better fit the tracing model. Once the processor is implemented (or closer to final PRs), they should be added to the semantic conventions in the spec. Not necessary to change it in this PR, can be changed later

File: processor/redactionprocessor/go.mod
Comment (dmitryax) [inline processor/redactionprocessor/go.mod:9]: Again, sorry for the delay. Please update the version to 0.38.0

**Review by dmitryax — APPROVED** [review]

File: processor/redactionprocessor/processor.go
Comment (leonsp-ai) [reply processor/redactionprocessor/processor.go:92]: @dmitryax Thanks for the link to the spec! I've pencilled in names that match a little better

File: processor/redactionprocessor/go.mod
Comment (leonsp-ai) [reply processor/redactionprocessor/go.mod]: Updated for 0.39.0

**Comment by leonsp-ai** [comment]
@codeboten What do you think about https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/5274#issuecomment-951677983 ?

Can we go ahead and merge regardless? cc @dmitryax @mx-psi

File: processor/redactionprocessor/README.md
Comment (dmitryax) [inline processor/redactionprocessor/README.md:65]: The new names look better please update them here as well

**Review by dmitryax — COMMENTED** [review]
The PR looks good to me. 

We need someone from @open-telemetry/collector-contrib-maintainer to merge this

**Comment by mx-psi** [comment]
@leonsp-ai can we try and add it to the `versions.yaml` list? https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/versions.yaml#L18 I think that may make the lint job pass

**Comment by jpkrohling** [comment]
Once the build passes, ping me and I'll merge this.

File: processor/redactionprocessor/README.md
Comment (leonsp-ai) [reply processor/redactionprocessor/README.md]: Those are metric names, so I'll move those to the follow-up metrics PR

**Comment by leonsp-ai** [comment]
@dmitryax Could you please approve the workflow? I've rebased which I think should fix the error in the last unittest/lint runs

File: processor/redactionprocessor/go.mod
Comment (jpkrohling) [inline processor/redactionprocessor/go.mod:8]: This needs to be in sync with what we have in the main branch at the moment of the merge. This is currently `v0.39.1-0.20211117203239-e23c9d0a0183`

Comment (leonsp-ai) [reply processor/redactionprocessor/go.mod]: @jpkrohling Updated to `v0.39.1-0.20211117203239-e23c9d0a0183`

Comment (jpkrohling) [reply processor/redactionprocessor/go.mod]: Sorry, looks like this got updated once again to v0.39.1-0.20211119172502-53d057f9c0e7 by #6390.

Comment (leonsp-ai) [reply processor/redactionprocessor/go.mod]: @jpkrohling Updated to `v0.39.1-0.20211119172502-53d057f9c0e7`

**Review by jpkrohling — APPROVED** [review]

**Pull request merged**
